Add MPL115A2 Barometric Pressure/Temperature Sensor

unic8s opened the pull request:
# What does this implement/fix?

Add support for the MPL115A2 pressure/temperature I²C sensor. This is the tinier version of the already exisiting implementation of the sensor MPL3115A2 but without the altitude metric. The code is partly ported from https://github.com/adafruit/Adafruit_MPL115A2

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3770

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
sensor:
- platform: mpl115a2
  temperature:
    name: "Temperature"
  pressure:
    name: "Pressure"
  update_interval: 10s
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

unic8s committed:
Initial renaming and quick port of Adafruits library

unic8s committed:
Fix variable names

unic8s committed:
Renamed in/out variables

unic8s committed:
Add coefficients on setup

unic8s committed:
Remove error status reading

unic8s committed:
Add into test yaml

unic8s committed:
Merge branch 'dev' into dev

unic8s committed:
Remove unused unit

unic8s committed:
Merge branch 'dev' of github.com:unic8s/esphome into dev

unic8s committed:
Removed whitespace

unic8s committed:
Add codeowners to python

unic8s committed:
Updated codeowners file

unic8s committed:
Fix black formatting

unic8s committed:
Fix code style

unic8s committed:
Fix variable naming

unic8s committed:
Merge branch 'dev' into dev

unic8s committed:
Merge branch 'dev' into dev

unic8s committed:
Merge branch 'dev' into dev

unic8s committed:
Merge branch 'dev' into dev

unic8s committed:
Merge branch 'dev' into dev

unic8s committed:
Merge branch 'dev' into dev

unic8s committed:
Merge branch 'dev' into dev

jesserockz commented on esphome/components/mpl115a2/mpl115a2.cpp:
```diff
@@ -0,0 +1,85 @@
+  this->write(&cmd, 1);
+  this->read(buffer, 8);
+
+  a0coeff = (((uint16_t) buffer[0] << 8) | buffer[1]);
+  b1coeff = (((uint16_t) buffer[2] << 8) | buffer[3]);
+  b2coeff = (((uint16_t) buffer[4] << 8) | buffer[5]);
+  c12coeff = (((uint16_t) buffer[6] << 8) | buffer[7]) >> 2;
```
```suggestion
  a0coeff = encode_uint16(buffer[0], buffer[1]);
  b1coeff = encode_uint16(buffer[2], buffer[3]);
  b2coeff = encode_uint16(buffer[4], buffer[5]);
  c12coeff = encode_uint16(buffer[6], buffer[7]) >> 2;
```

This will also require `#include "esphome/core/helpers.h"`

jesserockz commented on esphome/components/mpl115a2/mpl115a2.cpp:
```diff
@@ -0,0 +1,85 @@
+  b2coeff = (((uint16_t) buffer[4] << 8) | buffer[5]);
+  c12coeff = (((uint16_t) buffer[6] << 8) | buffer[7]) >> 2;
+
+  mpl115a2_a0_ = (float) a0coeff / 8;
+  mpl115a2_b1_ = (float) b1coeff / 8192;
+  mpl115a2_b2_ = (float) b2coeff / 16384;
+  mpl115a2_c12_ = (float) c12coeff;
+  mpl115a2_c12_ /= 4194304.0;
```
```suggestion
  this->mpl115a2_a0_ = a0coeff / 8.0f;
  this->mpl115a2_b1_ = b1coeff / 8192.0f;
  this->mpl115a2_b2_ = b2coeff / 16384.0f;
  this->mpl115a2_c12_ = c12coeff / 4194304.0f;
```

jesserockz commented on esphome/components/mpl115a2/mpl115a2.cpp:
```diff
@@ -0,0 +1,85 @@
+
+  this->write(cmd, 2);
+
+  // Wait a bit for the conversion to complete (3ms max)
+  delay(5);
+
+  cmd[0] = MPL115A2_REGISTER_PRESSURE_MSB;
+  this->write(cmd, 1);
+  this->read(buffer, 4);
+
+  pressure_in = (((uint16_t) buffer[0] << 8) | buffer[1]) >> 6;
+  temp_in = (((uint16_t) buffer[2] << 8) | buffer[3]) >> 6;
+
+  // See datasheet p.6 for evaluation sequence
+  pressure_comp = mpl115a2_a0_ + (mpl115a2_b1_ + mpl115a2_c12_ * temp_in) * pressure_in + mpl115a2_b2_ * temp_in;
+
+  float pressure_out = ((65.0F / 1023.0F) * pressure_comp) + 50.0F;
+  if (this->pressure_ != nullptr)
+    this->pressure_->publish_state(pressure_out);
+
+  float temperature_out = ((float) temp_in - 498.0F) / -5.35F + 25.0F;
+  if (this->temperature_ != nullptr)
+    this->temperature_->publish_state(temperature_out);
+
+  ESP_LOGD(TAG, "Got Temperature=%.1f°C Pressure=%.1f", temperature_out, pressure_out);
```
```suggestion
  // Wait a bit for the conversion to complete (3ms max)
  this->set_timeout(5, [this, cmd]() {
    uint8_t buffer[4];
	  cmd[0] = MPL115A2_REGISTER_PRESSURE_MSB;
	  this->write(cmd, 1);
	  this->read(buffer, 4);
	
	  uint16_t pressure_in = encode_uint16(buffer[0], buffer[1]) >> 6;
	  uint16_t temp_in = encode_uint16(buffer[2], buffer[3]) >> 6;
	
	  // See datasheet p.6 for evaluation sequence
	  float pressure_comp = mpl115a2_a0_ + (mpl115a2_b1_ + mpl115a2_c12_ * temp_in) * pressure_in + mpl115a2_b2_ * temp_in;
	
	  float pressure_out = ((65.0f / 1023.0f) * pressure_comp) + 50.0f;
	  if (this->pressure_ != nullptr)
	    this->pressure_->publish_state(pressure_out);
	
	  float temperature_out = ((float) temp_in - 498.0f) / -5.35f + 25.0f;
	  if (this->temperature_ != nullptr)
	    this->temperature_->publish_state(temperature_out);
	
	  ESP_LOGD(TAG, "Got Temperature=%.1f°C Pressure=%.1f", temperature_out, pressure_out);
  });
```

jesserockz commented on esphome/components/mpl115a2/mpl115a2.cpp:
```diff
@@ -0,0 +1,85 @@
+    this->temperature_->publish_state(temperature_out);
+
+  ESP_LOGD(TAG, "Got Temperature=%.1f°C Pressure=%.1f", temperature_out, pressure_out);
+
+  this->status_clear_warning();
```
There is nowhere a warning is set, so no need to clear
```suggestion
```

jesserockz commented on esphome/components/mpl115a2/mpl115a2.h:
```diff
@@ -0,0 +1,49 @@
+
+class MPL115A2Component : public PollingComponent, public i2c::I2CDevice {
+ public:
+  void set_temperature(sensor::Sensor *temperature) { temperature_ = temperature; }
+  void set_pressure(sensor::Sensor *pressure) { pressure_ = pressure; }
```
```suggestion
  void set_temperature(sensor::Sensor *temperature) { this->temperature_ = temperature; }
  void set_pressure(sensor::Sensor *pressure) { this->pressure_ = pressure; }
```

jesserockz commented on esphome/components/mpl115a2/mpl115a2.cpp:
```diff
@@ -0,0 +1,85 @@
+}
+
+void MPL115A2Component::update() {
+  uint16_t pressure_in, temp_in;
+  float pressure_comp;
+
+  uint8_t cmd[2] = {MPL115A2_REGISTER_STARTCONVERSION, 0};
+  uint8_t buffer[4];
+
```
```suggestion
  uint8_t cmd[2] = {MPL115A2_REGISTER_STARTCONVERSION, 0};
```

jesserockz commented on esphome/components/mpl115a2/sensor.py:
```diff
@@ -0,0 +1,57 @@
+    await cg.register_component(var, config)
+    await i2c.register_i2c_device(var, config)
+
+    if CONF_PRESSURE in config:
+        sens = await sensor.new_sensor(config[CONF_PRESSURE])
```
```suggestion
    if pressure_sensor := config.get(CONF_PRESSURE):
        sens = await sensor.new_sensor(pressure_sensor)
```

jesserockz commented on esphome/components/mpl115a2/sensor.py:
```diff
@@ -0,0 +1,57 @@
+        sens = await sensor.new_sensor(config[CONF_PRESSURE])
+        cg.add(var.set_pressure(sens))
+
+    if CONF_TEMPERATURE in config:
+        sens = await sensor.new_sensor(config[CONF_TEMPERATURE])
```
```suggestion
    if temperature_sensor := config.get(CONF_TEMPERATURE):
        sens = await sensor.new_sensor(temperature_sensor)
```

jesserockz commented on tests/test1.yaml:
Please move this to its own file in `tests/components/mpl115a2/`

unic8s replied:
@jesserockz I did so - should I remove it from the `test1.yaml` as well?

unic8s committed:
Apply suggestions from code review

@jesserockz thank you very much for your review and helping me out on the topics I missed

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

unic8s committed:
Added component tests

unic8s committed:
Fixed formatting

unic8s committed:
Fix by moving the cmd inside the timeout

unic8s committed:
Merge branch 'dev' into dev

unic8s committed:
Merge branch 'esphome:dev' into dev

unic8s reviewed (commented):
Looks good to me. I did some changes to fix the reference issue inside the timeout callback. Thank you very much

unic8s committed:
Merge branch 'dev' into dev
